zcash_client_sqlite/src/lib.rs

```diff
@@ -1049,19 +1050,28 @@ impl<P: consensus::Parameters> WalletWrite for WalletDb<rusqlite::Connection, P>
     ) -> Result<(), Self::Error> {
         self.transactionally(|wdb| {
             let tx_ref = wallet::put_tx_data(wdb.conn.0, d_tx.tx(), None, None)?;
+            let funding_accounts = wallet::get_funding_accounts(wdb.conn.0, d_tx.tx())?;
+            let funding_account = funding_accounts.iter().next().copied();
+            if funding_accounts.len() > 1 {
+                warn!(
+                    "More than one wallet account detected as funding transaction {:?}, selecting {:?}",
+                    d_tx.tx().txid(),
+                    funding_account.unwrap()
+                )
+            }
 
-            let mut spending_account_id: Option<AccountId> = None;
             for output in d_tx.sapling_outputs() {
                 match output.transfer_type() {
                     TransferType::Outgoing | TransferType::WalletInternal => {
                         let recipient = if output.transfer_type() == TransferType::Outgoing {
                             //TODO: Recover the UA, if possible.
                             Recipient::Sapling(output.note().recipient())
                         } else {
-                            Recipient::InternalAccount(
-                                *output.account(),
-                                Note::Sapling(output.note().clone()),
-                            )
+                            Recipient::InternalAccount {
+                                receiving_account: *output.account(),
+                                external_address: None,
```

Comment:
How can we ensure this is set correctly for cross-account transfers instead of setting it to `None`?

Reply:
What I mean is, don't we need a separate `TransferType` case to distinguish between wallet-internal (cross-account) and account-internal?

Reply:
Oh, I guess that for correct cross-account transfers (i.e. correctly treated as separate accounts, so that each account's external FVK sees the correct view), they should show up here as `TransferType::Outgoing`. In that case, we should have `external_address` available already for cross-account transfers and can set it here?

Reply:
Ah, and now going back into the PR diff, we're currently in the `TransferType::WalletInternal` branch. Okay, so the change I think is missing here is that the other side of this branch should be tracking the receiving account for cross-account transfers.

Reply:
Discussion from offline: the received_notes tables would then duplicate exactly information that's available from the sent_notes table, and such duplication would be a potential source of error. So this should be fine.

Reply:
When we scan, we scan with _all_ of the available IVKs, so in the case of a cross-account transfer it will always be detected as `TransferType::Incoming` and will get handled by that branch, which will correctly find and set the source account. And the `Outgoing` case is already handled correctly.

Reply:
There's nothing that guarantees that (from the perspective of an implementation of the `WalletWrite` trait), but I agree that's what the current code does in `zcash_client_backend` which is what callers are most likely using, as I said in my subsequent review. It's just brittle.